AkitaInteractive/Hachiko-Engine
# Saving and Loading Scripts Via Code Generation

# What's new:

With this PR we have the following functionalities for scripting:
1. Built in `math::Quat` support for editor.
2. Yaml serialization code generation for scripts. Now with this, the variables you set from editor will be saved when you save the scene, and you won't have to set those variables on `Script::OnAwake`, including all `Component*` and `GameObject*`!

Please note that, we still don't support `Script*` serialization both in runtime and on load-save time. But still I have the solution on mind, just didn't had the time as this is not a priority, as you can store GameObject* instead and drag and drop it, and on `Script::OnAwake` you can do `non_serialized_script_ptr_field = game_object_with_script->GetComponent<ScriptType>();`.

Also note that only things you see on editor are saved at save-load time, but many other things that doesn't have editor support are still have the ability to be serialized at runtime.

# How To Test:

1. Build hachiko, and then build Gameplay, Be sure to check file contents on Gameplay/generated/Factory.cpp,  Gameplay/generated/SerializationMethods.cpp, Gameplay/generated/GeneratedEditor.cpp and SaveAndLoadMethods.cpp are changed after you run the build.
2. Open an empty scene, add GameObjects, components and scripts and change the script fields and save the scene.
3. Play the scene and see the fields. If in the script those fields are changed on `Script::OnAwake` they should be overriden by it.
4. Stop the scene and see that the fields are just as your latest save.

THIS PR MARKS v3.0 of Chad++ Scripting System™.
